Interaction functions

Gengrui-Zhang opened the pull request:
Hi Mark, 

I have cleaned my branch and merged the most updated master branch into the interaction branch. 
The function for creating product indicators with se and loading is named "get_fs_int()". A vignette with the demonstration is also included. 
Since I want to keep the most relevant information for interactions, I deleted some irrelevant files (but you can let me know if it's inappropriate). The CMD check gave 0 errors, 1 warning, and 3 notes. 
I have moved many files to a new repo called "2S-PA-Int", and I will develop more interaction-related stuff there. I will invite you to that repo as well, so you can leave some comments. 
Also, I would like to simplify and improve my other functions, and also make accommodations given the changes to `get_fs()` and `tspa()`

Thanks!
Jimmy

marklhc committed:
Merge remote-tracking branch 'origin/covariances' into 27-2spa-for-latent-interactions

marklhc committed:
Merge branch 'covariances' into 27-2spa-for-latent-interactions

marklhc committed:
Create new vignette for interaction #27

Gengrui-Zhang committed:
update

Gengrui-Zhang committed:
update

marklhc committed:
Fix typo

marklhc committed:
Merge branch '27-2spa-for-latent-interactions' of https://github.com/Gengrui-Zhang/R2spa into 27-2spa-for-latent-interactions

# Conflicts:
#	vignettes/tspa-interaction.Rmd

Gengrui-Zhang committed:
Update tspa function with interaction, and unconstrained product indicator method

Gengrui-Zhang committed:
commit

Merge branch '27-2spa-for-latent-interactions' of https://github.com/Gengrui-Zhang/R2spa into 27-2spa-for-latent-interactions

# Conflicts:
#	vignettes/tspa-interaction.Rmd

Gengrui-Zhang committed:
Add functions to vignette

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
Separate functions

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit changes

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
update simulation results

Gengrui-Zhang committed:
update simulation results

Gengrui-Zhang committed:
commit name change

Gengrui-Zhang committed:
commit changes

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
run high covariance and high reliability condition

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
Commit

Gengrui-Zhang committed:
error constraints

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
Temporary commit

Gengrui-Zhang committed:
Revised `rapi()`

Gengrui-Zhang committed:
update rapi function and vignette

Gengrui-Zhang committed:
Update scripts and simulation results

Gengrui-Zhang committed:
Update paper draft and simulation script

Gengrui-Zhang committed:
update NAMESPACE

Gengrui-Zhang committed:
Commit the paper draft and final version of results

Gengrui-Zhang committed:
Rename paper draft

Gengrui-Zhang committed:
commit new materials

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
Add interaction function

Gengrui-Zhang committed:
rename interaction function

marklhc commented:
Thanks @Gengrui-Zhang. Could you change the destination of merging from `master` to `devel`? Also, could you exclude files under the  "doc" folder from the PR? Thank you.

Gengrui-Zhang commented:
> Thanks @Gengrui-Zhang. Could you change the destination of merging from `master` to `devel`? Also, could you exclude files under the "doc" folder from the PR? Thank you.

Hi @marklhc , I have changed the destination to `devel`. As for the "doc" folder, it seems that it does not appear in the `devel` branch. Do you mean by manually restoring them in my branch?

marklhc commented:
@Gengrui-Zhang , it seems that you have deleted many of the files, including the vignettes, in 2fe766cd, and then it looks like you manually restored them in e385367. This is going to cause the file history of those files all gone once we merge things, so I cannot let that happen. I may need to rebase your branch to [2de9e17](https://github.com/Gengrui-Zhang/R2spa/pull/82/commits/2de9e1723d5f3ced20951cabda26c921257136d6) and then we go from there.

marklhc commented:
And for some reason it looks like you didn't do a merge properly in [2de9e17](https://github.com/Gengrui-Zhang/R2spa/pull/82/commits/2de9e1723d5f3ced20951cabda26c921257136d6). I will need to try again

marklhc commented:
And please don't manually copy code from other branches to your branch; use a merge commit and resolve conflicts. Please make sure you know how to properly do a merge commit.

marklhc committed:
Remove unneeded files

marklhc committed:
Merge branch 'devel' into 27-2spa-for-latent-interactions

marklhc committed:
Delete documentation for renamed function

marklhc committed:
Add documentation for renamed file

marklhc committed:
Delete unncessary files

marklhc commented on R/tspa.R:
```diff
@@ -1,3 +1,5 @@
+# Interaction-branch
```
Please delete the extra words

marklhc commented on R/tspa.R:
```diff
@@ -28,7 +30,7 @@
 #' @param ... Additional arguments passed to \code{\link[lavaan]{sem}}. See
 #'            \code{\link[lavaan]{lavOptions}} for a complete list.
-#' @return An object of class \code{lavaan}, with an attribute \code{tspaModel}
+#' @return An object of classy7777777777777777777777777777777777777y \code{lavaan}, with an attribute \code{tspaModel}
```
Please revert to the original version.

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,61 @@
+#'         with their loadings and standard errors.
+#'
+#' @importFrom utils combn
+#'
```
Please provide an example for the function.

marklhc replied:
I didn't see an example added in the documentation

marklhc committed:
Remove unncessary changes

marklhc committed:
Basic cleanup

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+  dat_pi <- data.frame(matrix(ncol = 2*length(fs_pairs), nrow = nrow(dat)))
+  for(i in seq_along(fs_pairs)) {
+    pair <- fs_pairs[[i]]
+    dat_pi[,2*i-1] <- fs_dat[[pair[1]]]*fs_dat[[pair[2]]] - mean(fs_dat[[pair[1]]]*fs_dat[[pair[2]]])
```
`fs_dat` in not defined. Please revise your function.

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+#'
+#' @export
+
+get_fs_int <- function(dat, fs_name, se, loading, model = NULL) {
```
To be consistent with the arguments in `tspa()`

```suggestion
get_fs_int <- function(dat, fs_name, se_fs, loading_fs, model = NULL) {
```

marklhc replied:
For now, we can force users to have `se_fs` and `loading_fs` to be columns in `dat`, and make sure that the inputs are character vectors.

Gengrui-Zhang replied:
Got it. Thanks!

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+#' @export
+
+get_fs_int <- function(dat, fs_name, se, loading, model = NULL) {
+
```
@Gengrui-Zhang, please add an error when the input types are not correct, or when `dat` does not contain colum names that match the input data.

Gengrui-Zhang replied:
I added a bunch of error-check commands. Please let me know what you think. Thanks!

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+  if (is.null(model)) {
+    fs_pairs <- combn(fs_name, 2, simplify = FALSE)
+  } else {
+    elements <- trimws(unlist(strsplit(model, "\\+")))
+    fs_pairs <- lapply(strsplit(elements, ":"), function(pair) {
+      sapply(pair, function(element) {
+        paste0(names(element), element)
+      })})
```
```suggestion
    fs_pairs <- lapply(unlist(strsplit(model, split = "\\+")),
                       FUN = function(pair) { 
                         pair_nospace <- trimws(pair)
                         unlist(strsplit(pair_nospace, split = ":"))
                       })
```

Gengrui-Zhang replied:
Great suggestion. Thanks!

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+      })})
+  }
+
+  # Observation-specific check
+  check_element <- function(x) {
+    if(var(x, na.rm = TRUE) == 0) {
+      return(unlist(x[1]))
+    } else {
+      return(x)
+    }
+  }
```
Not needed with the change in input type

```suggestion

```

Gengrui-Zhang replied:
Removed

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+                     se_2 = unname(unlist(fs_dat[fs_list[[pair[2]]]$se])))
+    par_list <- lapply(par_list, check_element)
+
+    dat_pi[,2*i] <- as.data.frame(matrix(rep(sqrt(par_list$loading_2^2*par_list$se_1^2 +
```
Right now you're assuming that the latent variances are 1, but this may not be the case. Should you allow users to specify the latent variances too?

Gengrui-Zhang replied:
I have added an argument `lat_var` to allow for user-specified latent variances. The default values are 1(s) for all latent variables

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  # Create PI with SE
+  dat_pi <- data.frame(matrix(ncol = 2*length(fs_pairs), nrow = nrow(dat)))
+  for(i in seq_along(fs_pairs)) {
+    pair <- fs_pairs[[i]]
+    dat_pi[,2*i-1] <- fs_dat[[pair[1]]]*fs_dat[[pair[2]]] - mean(fs_dat[[pair[1]]]*fs_dat[[pair[2]]])
+
+    par_list <- list(loading_1 = unname(unlist(fs_dat[fs_list[[pair[1]]]$loading])),
+                     loading_2 = unname(unlist(fs_dat[fs_list[[pair[2]]]$loading])),
+                     se_1 = unname(unlist(fs_dat[fs_list[[pair[1]]]$se])),
+                     se_2 = unname(unlist(fs_dat[fs_list[[pair[2]]]$se])))
+    par_list <- lapply(par_list, check_element)
+
+    dat_pi[,2*i] <- as.data.frame(matrix(rep(sqrt(par_list$loading_2^2*par_list$se_1^2 +
+                                                    par_list$loading_1^2*par_list$se_2^2 +
+                                                    par_list$se_1^2*par_list$se_2^2),
+                                             length(dat_pi[,i]))))
+    colnames(dat_pi)[2*i-1] <- paste(pair[1], pair[2], sep = ".")
+    colnames(dat_pi)[2*i] <- paste(pair[1], pair[2], "se", sep = ".")
+  }
```
```suggestion
dat_pi <- dat
  for (i in seq_along(fs_pairs)) {
    pair <- fs_pairs[[i]]
    name_i <- paste(pair, collapse = ":")
    dat_pi[[name_i]] <- dat[[pair[1]]] * dat[[pair[2]]] -
      mean(dat[[pair[1]]] * dat[[pair[2]]])
    name_se_i <- paste0(name_i, "_se")
    name_ld_i <- paste0(name_i, "_ld")
    se_vars <- se_fs[match(pair, table = fs_name)]
    loading_vars <- loading_fs[match(pair, table = fs_name)]
    dat_pi[[name_se_i]] <- sqrt(
      dat[[loading_vars[1]]]^2 * dat[[se_vars[2]]]^2 +
      dat[[loading_vars[2]]]^2 * dat[[se_vars[1]]]^2 +
      dat[[se_vars[1]]]^2 * dat[[se_vars[2]]]^2
    )
    # Also loadings
    dat_pi[[name_ld_i]] <- dat[[loading_vars[1]]] * dat[[loading_vars[2]]]
  }
```

Gengrui-Zhang replied:
Thanks for providing your structure of coding. This looks more organized and cleaner!

Gengrui-Zhang committed:
Commit:
1. Corrected names of function in the vignette
2. Added input-check function
3. changed names for se and loadings to match the tspa() function

Gengrui-Zhang committed:
More informative code for creating fs pairs

Gengrui-Zhang committed:
1. Added one example with specified latent variance
2. Changed the main function with Mark's suggestion.

Gengrui-Zhang committed:
Add a test file for `get_fs_int()`

Gengrui-Zhang commented:
Hii @marklhc, thanks so much for your inspiring comments and practical suggestions. I have addressed all the comments and added test file `test-get_fs_int()` for this function. Please let me know if you have any questions or further suggestions/comments. Thanks so much!

marklhc committed:
quick commit

marklhc committed:
Merge remote-tracking branch 'origin/27-2spa-for-latent-interactions' into 27-2spa-for-latent-interactions

marklhc committed:
Clean up

marklhc commented on vignettes/get_fs_int-vignette.Rmd:
````diff
@@ -0,0 +1,115 @@
+
+```{r load packages}
+library(lavaan)
+library(dplyr)
````
For package development, suggest to use base packages if possible. 

```suggestion

```

marklhc committed:
Clean up vignette

marklhc committed:
Add example in documentation

marklhc committed:
Fix a new bug in the `se_fs` argument in `tspa()`

marklhc commented on vignettes/get_fs_int-vignette.Rmd:
```diff
@@ -0,0 +1,125 @@
+                              m =~ m1 + m2 + m3
+                              z =~ z1 + z2 + z3
+                              y =~ y1 + y2 + y3",
+                 method = "Bartlett")
```
You should specify the `std.lv = TRUE` option

Gengrui-Zhang replied:
This is added.

marklhc commented on vignettes/get_fs_int-vignette.Rmd:
````diff
@@ -0,0 +1,125 @@
+
+```{r Demonstrating int_ind() function}
+# With specified model
+ind_1 <- get_fs_int(dat = fs_dat,
````
The code is only valid when the factor variances are assumed to be one.

Gengrui-Zhang replied:
From line 120, I clarified that "The above example assumes that the factor scores are obtained from a model where the latent variances are assumed to be 1", and a third example specifying latent variances is added below (i.e., `lat_var = lat_var`). Is this OK?

marklhc replied:
Thanks

marklhc commented on R/get_fs_int.R:
```diff
@@ -0,0 +1,80 @@
+#' Compute interaction indicators for tspa() function
```
It should be clarified that double-mean-centering is used. Documentation should include more technical details!

Gengrui-Zhang replied:
A sentence for DMC is added

marklhc committed:
Further clean-up documentation

marklhc committed:
Elaborate more in vignette

Gengrui-Zhang committed:
Added a description of DMC strategy

marklhc commented on vignettes/get_fs_int-vignette.Rmd:
````diff
@@ -0,0 +1,146 @@
+We will simulate a dataset with four firsr-order latent variables: x, m, z, y
+
+```{r Data Simulation}
+## Sample size:
````
I dont think you can (or should) use space in chunk label

marklhc committed:
Add subsequent analysis to vignette

marklhc committed:
Use a different source of the data for growth vignette
